### ISSUE TITLE
Problem with proper certificate chain verification in TLS connection

### ISSUE THREAD

**Issue by thebodzio** [lavv17/lftp#143]
I've got a problem with proper certificate chain verification in TLS connection. I'm connecting with my server using FTPS explicitly, but the very same thing happens when connection is made via plain FTP and lftp tries to authenticate with TLS. Here's an excerpt from lftp output in debug mode:

```
Certificate: C=PL,OU=Domain Control Validated,CN=ruthenus.pl      
 Issued by:        C=BE,O=GlobalSign nv-sa,CN=AlphaSSL CA - SHA256 - G2
 Checking against: C=BE,O=GlobalSign nv-sa,CN=AlphaSSL CA - SHA256 - G2
  Trusted
Certificate: C=BE,O=GlobalSign nv-sa,CN=AlphaSSL CA - SHA256 - G2
 Issued by:        C=BE,O=GlobalSign nv-sa,OU=Root CA,CN=GlobalSign Root CA
 Checking against: O=AlphaSSL,CN=AlphaSSL CA - G2
ERROR: Certificate verification: Not trusted: no issuer was found
Certificate: O=AlphaSSL,CN=AlphaSSL CA - G2
 Issued by:        C=BE,O=GlobalSign nv-sa,OU=Root CA,CN=GlobalSign Root CA
 Checking against: C=BE,O=GlobalSign nv-sa,CN=GlobalSign Domain Validation CA - SHA256 - G2
ERROR: Certificate verification: Not trusted: no issuer was found
Certificate: C=BE,O=GlobalSign nv-sa,CN=GlobalSign Domain Validation CA - SHA256 - G2
 Issued by: OU=GlobalSign Root CA - R3,O=GlobalSign,CN=GlobalSign
  Trusted
**** Certificate verification: Not trusted: no issuer was found
```

Connections to the same server made with `openssl` and `gnutls-cli` were made successfully—in both cases certificates validated as expected.

I've also tried to point lftp to proper CA bundle with `ssl:ca-file`, but, sadly, to no avail.

lftp is built with GnuTLS (basic Arch Linux issued package from extra repo). Here's `ldd` output, for reference:

```
	linux-vdso.so.1 (0x00007ffc3f3d6000)
	libexpat.so.1 => /usr/lib/libexpat.so.1 (0x00007f60998f3000)
	libz.so.1 => /usr/lib/libz.so.1 (0x00007f60996dd000)
	libgnutls.so.28 => /usr/lib/libgnutls.so.28 (0x00007f60993ba000)
	libreadline.so.6 => /usr/lib/libreadline.so.6 (0x00007f609916f000)
	libutil.so.1 => /usr/lib/libutil.so.1 (0x00007f6098f6c000)
	libncursesw.so.5 => /usr/lib/libncursesw.so.5 (0x00007f6098d07000)
	libdl.so.2 => /usr/lib/libdl.so.2 (0x00007f6098b03000)
	libidn.so.11 => /usr/lib/libidn.so.11 (0x00007f60988cf000)
	libstdc++.so.6 => /usr/lib/libstdc++.so.6 (0x00007f60985c0000)
	libc.so.6 => /usr/lib/libc.so.6 (0x00007f609821d000)
	libp11-kit.so.0 => /usr/lib/libp11-kit.so.0 (0x00007f6097fb7000)
	libtasn1.so.6 => /usr/lib/libtasn1.so.6 (0x00007f6097da4000)
	libnettle.so.4 => /usr/lib/libnettle.so.4 (0x00007f6097b76000)
	libhogweed.so.2 => /usr/lib/libhogweed.so.2 (0x00007f6097947000)
	libgmp.so.10 => /usr/lib/libgmp.so.10 (0x00007f60976b5000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f6099b1d000)
	libm.so.6 => /usr/lib/libm.so.6 (0x00007f60973b0000)
	libgcc_s.so.1 => /usr/lib/libgcc_s.so.1 (0x00007f609719a000)
	libffi.so.6 => /usr/lib/libffi.so.6 (0x00007f6096f91000)
	libpthread.so.0 => /usr/lib/libpthread.so.0 (0x00007f6096d74000)
```

I'm sorry, but I just run out of ideas what to try next :}

lftp 4.6.1
gnutls 3.3.14

**Comment by lavv17** [comment]
It looks like a bug in gnutls function gnutls_certificate_get_peers. A
certificate seems to be missing in the chain returned by that function.

пт, 17 апр. 2015, 3:33, thebodzio <notifications@github.com>:

**Comment by mruprich** [comment]
It says in the gnutls API reference: "the server must send as first certificate in the list its own certificate, following the issuer's certificate, then the issuer's issuer etc. However, there are servers which violate this principle and thus on certain occasions this may be an unsorted list." Is there a possibility that this could be the case and lftp is unable to deal with the unsorted list? (just an idea)

**Comment by thebodzio** [comment]
I'm glad to see somebody is interested in this issue.

Unfortunately I don't think the improper order of certificates is the case. The order of certificates mentioned in debug log in my initial message is exactly as it should be (own cert, issuer, issuer's issuer and so on). The problem is for some reason gnutls doesn't validate the second and the following certificates. That is, of course, as long as debug log lists the certificates in the same order as they're placed in the server's response.

I think perhaps checking gnutls-cli code could shed some light here, because I could establish connection with that program just fine.

**Comment by mruprich** [comment]
Yes I tried this with gnutls-cli as well and no problem there. I might try to look at the gnutls-cli code and see what is going on there.

**Comment by thebodzio** [comment]
Thank you! Your effort is most appreciated.

**Comment by GenaBitu** [comment]
Hi, I have been plagued by this problem as well... Can I do anything to help?

**Comment by colinkelly** [comment]
Did anyone got further in the troubleshooting for this issue?

On my side, I see that debian based machine do not give out the error however, rhel based do have this error.

**Comment by MayeulC** [comment]
I have the same issue, [here is my SuperUser question](https://superuser.com/questions/1414353/trust-server-certificate-with-lftp) about it.

For what it's worth, FileZilla seems to agree with @mruprich:
> Server sent unsorted certificate chain in violation of the TLS specifications
